node/src/storage_mngr/node_migrations.rs

```diff
@@ -0,0 +1,296 @@
+            "Error when reading ChainState from database: version {} not supported",
+            unknown_version
+        )),
+        Err(()) => {
+            // Error reading version (end of file?)
+            Err(failure::format_err!(
+                "Error when reading ChainState version from database: unexpected end of file"
+            ))
+        }
+    }
+}
+
+/// Get value associated to key, with migrations support
+fn get_versioned<K, V, F>(
+    key: &K,
+    migration_fn: F,
+) -> impl Future<Output = Result<Option<V>, failure::Error>>
+where
+    K: serde::Serialize,
+    F: FnOnce(Vec<u8>) -> Result<V, failure::Error>,
+{
+    let addr = StorageManagerAdapter::from_registry();
+
+    let key_bytes = match serialize(key) {
+        Ok(x) => x,
+        Err(e) => return futures::future::Either::Left(future::ready(Err(e.into()))),
+    };
+
+    let fut = async move {
+        let opt = addr.send(Get(key_bytes)).await??;
+
+        match opt {
+            Some(bytes) => migration_fn(bytes).map(Some),
+            None => Ok(None),
+        }
+    };
+
+    futures::future::Either::Right(fut)
+}
+
+/// Get value associated to key
+pub fn get_chain_state<K>(
+    key: &K,
+) -> impl Future<Output = Result<Option<ChainState>, failure::Error>>
+where
+    K: serde::Serialize,
+{
+    get_versioned(key, |bytes| migrate_chain_state(&bytes))
+}
+
+/// Put a value associated to the key into the storage, preceded by a 4-byte version tag
+fn put_versioned<'a, 'b, K>(
+    key: &'a K,
+    value: &'b ChainState,
+    db_version: u32,
+) -> impl Future<Output = Result<(), failure::Error>> + 'static
+where
+    K: serde::Serialize,
+{
+    let addr = StorageManagerAdapter::from_registry();
+
+    let key_bytes = match serialize(key) {
+        Ok(x) => x,
+        Err(e) => {
+            return futures::future::Either::Left(futures::future::Either::Right(future::ready(
+                Err(e.into()),
+            )))
+        }
+    };
+
+    let mut buf = db_version.to_le_bytes().to_vec();
+    let value_bytes = match bincode::serialize_into(&mut buf, value) {
+        Ok(()) => buf,
+        Err(e) => {
+            return futures::future::Either::Left(futures::future::Either::Left(future::ready(
+                Err(e.into()),
+            )))
+        }
+    };
+
+    futures::future::Either::Right(async move { addr.send(Put(key_bytes, value_bytes)).await? })
+}
+
+/// Put a value associated to the key into the storage
+// TODO: how to ensure that we don't accidentally persist the chain state using put instead of put_chain_state?
+pub fn put_chain_state<'a, 'b, K>(
+    key: &'a K,
+    chain_state: &'b ChainState,
+) -> impl Future<Output = Result<(), failure::Error>> + 'static
+where
+    K: serde::Serialize + 'static,
+{
+    let db_version: u32 = 2;
+    // The first byte of the ChainState db_version must never be 0 or 1,
+    // because that can be confused with version 0.
+    assert!(db_version.to_le_bytes()[0] >= 2);
```

Comment:
@aesedepece as long as this assert is not lost during some refactor, any attempts to set the version to an invalid number will fail.

Reply:
Nice!